sys/vfs_readdir.c

```diff
@@ -0,0 +1,41 @@
+#include <common.h>
+#include <dirent.h>
+#include <malloc.h>
+#include <uio.h>
+#include <vfs.h>
+
+int readdir_generic(vnode_t *v, uio_t *uio, readdir_ops_t *ops) {
+  void *it = ops->first(v);
+  dirent_t *dir;
+  off_t offset = 0;
+  int error;
+
+  /* Locate proper directory based on offset */
+  for (; it; it = ops->next(it)) {
+    unsigned reclen = _DIRENT_RECLEN(dir, ops->namlen_of(it));
+    if (offset + reclen > uio->uio_offset) {
+      assert(it == NULL || offset == uio->uio_offset);
+      break;
+    }
+    offset += reclen;
+  }
```

Comment:
[VOP_READDIR(9)](http://netbsd.gw.com/cgi-bin/man-cgi?VOP_READDIR) states following:
 
>The directory contents are read into `struct dirent` structures and `uio->uio_offset` is set to the offset of the next unread directory entry.  This offset may be used in a following invocation to continue a sequential read of the directory contents.

In all contexts I know `uio_offset` value is expressed in bytes. I would prefer not change semantics of this field to anything else, just to avoid any suprises for future readers. Moreover value returned in `basep` is allowed to be passed to `lseek`, which again accepts offset in bytes, not records.

If I were to optimize directory traversing I'd probably keep `void *entry` somewhere in `file_t` and extend `VOP_READDIR` with an extra argument to pass the state of `readdir_generic`, so consecutive syscalls reuse its value. In fact there's some provision in NetBSD interface, which seems to enable some optimizations, but they do it somewhat differently:

```c
int VOP_READDIR(struct vnode *vp, struct uio *uio, kauth_cred_t cred, int *eofflag,
                off_t **cookies, int *ncookies);
```

> In case `ncookies` and `cookies` are supplied, one cookie should be returned per directory entry.  The value of the cookie for each directory entry should be the offset within the directory where the on-disk version of the following directory entry starts. That is, for each directory entry i, the corresponding cookie should refer to the offset of directory entry i + 1.